provider/pkg/provider/handlers.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021, Pulumi Corporation.
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-google-native/provider/pkg/resources"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+)
+
+type updateHandler interface {
+	Update(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error)
+}
+
+type updateHandlerFunc func(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error)
+
+func (u updateHandlerFunc) Update(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error) {
+	return u(inputs, oldState)
+}
+
+func (p *googleCloudProvider) updateHandler(resourceTok string, req *rpc.UpdateRequest) updateHandler {
+	logging.V(9).Infof("UpdateHandler for %q", resourceTok)
+	switch resourceTok {
+	case "google-native:container/v1:Cluster", "google-native:container/v1beta1:Cluster":
+		return updateHandlerFunc(func(inputs, oldState resource.PropertyMap) (map[string]interface{}, error) {
+			res, ok := p.resourceMap.Resources[resourceTok]
+			if !ok {
+				return nil, errors.Errorf("resource %q not found", resourceTok)
+			}
+
+			updateBody := p.prepareAPIInputs(inputs, oldState, res.UpdateProperties)
+
+			// TODO: Cluster updates also include flattened members of the `networkConfig` map
+			// which is hard to represent with the current metadata format in place (sdkName
+			// can only be used to refer to fields at the top level).
+
+			var splitBody []map[string]interface{}
+			update, hasUpdate := updateBody["update"]
+			updateMap, isMap := update.(map[string]interface{})
+			if hasUpdate && isMap {
+				for k, v := range updateMap {
+					if strings.HasPrefix(k, "desired") {
+						splitBody = append(splitBody, map[string]interface{}{k: v})
+					}
+				}
+			}
+
+			logging.V(9).Infof("updateBody: %+v", updateBody)
+			logging.V(9).Infof("splitBody: %+v", splitBody)
+
+			// Update operation on GKE clusters are to be performed one field at a time.
```

Comment:
The terraform provider performs some [cluster level locking](https://github.com/hashicorp/terraform-provider-google-beta/blob/master/google-beta/resource_container_cluster.go#L1861) to prevent concurrent updates to the same cluster. Is this something we need to guard against as well?

Reply:
What would be a scenario when we'd get a concurrent update?